### DISCUSSION TITLE
How to point to a local language server?

### DISCUSSION THREAD

**Discussion by olrtg** [LunarVim/LunarVim#4012] [Q&A] [answered]
Hey guys! I want to setup a language server that I'm making a contribution to (`emmet-ls`).. I have this language server built with my changes and it is in my `$PATH`.. but I don't know how to point to my version of the language server instead of the mason one.. any ideas on how to achieve this?

**Comment by LostInTheLogs** [comment]
uninstall the mason one or prepend vim.env.PATH with your lang server(I'm not sure if this one will work though)

**Reply by olrtg** [reply, marked as answer]
Well I found that you can use a language server in your path by using `require("lspconfig").emmet_ls.setup({...})` instead of the one provided by LunarVim `require("lvim.lsp.manager").setup("emmet_ls", {...})`

**Comment by simpers** [comment]
Would it be possible though to _only_ override the location of the language server, and keep the rest of the settings as LunarVim/mason defaults?

**Reply by LostInTheLogs** [reply]
use this example from the docs and instead of changing `settings` change `cmd` https://www.lunarvim.org/docs/configuration/language-features/language-servers#by-overriding-the-setup

**Reply by simpers** [reply]
Does this require it to be _installed_? Meaning I'd have to install it but then override it with my own version? Or will this count it as installed and proceed?

And if I do provide settings, are the defaults still respected? Meaning my settings map will be merged with the default one?

**Reply by LostInTheLogs** [reply]
>  Does this require it to be installed? Meaning I'd have to install it but then override it with my own version? Or will this count it as installed and proceed?

if it isn't in $PATH then it will need to be installed also in mason, if it is in $PATH then just uninstall the mason version and it should work

> And if I do provide settings, are the defaults still respected? Meaning my settings map will be merged with the default one?

your setting will be merged with the defaults

**Reply by LostInTheLogs** [reply]
if you're using automatic installation (on by default) you'll also need to add the server to the `lvim.lsp.installer.setup.automatic_installation.exclude` table

**Reply by simpers** [reply]
> if you're using automatic installation (on by default) you'll also need to add the server to the `lvim.lsp.installer.setup.automatic_installation.exclude` table

This I have done! ✅ 

> > Does this require it to be installed? Meaning I'd have to install it but then override it with my own version? Or will this count it as installed and proceed?
> 
> if it isn't in $PATH then it will need to be installed also in mason, if it is in $PATH then just uninstall the mason version and it should work
> 
> > And if I do provide settings, are the defaults still respected? Meaning my settings map will be merged with the default one?
> 
> your setting will be merged with the defaults

Great about the settings ✅ 

About the `$PATH` though; does it expect `language_server.sh` to be available in `$PATH` or the command that I supplied it? Meaning if I make a symlink using `ln -s elixir-ls language-server.sh`, assuming that would work as expected with the server itself, then LunarVim will be able to run `elixir-ls` if I set that as the cmd?

Right now I have set cmd like this:

```lua
local path_to_elixirls = vim.fn.expand("~/.local/elixir-ls-0.14.5/language_server.sh")
require("mason-lspconfig").setup_handlers {
  ["elixirls"] = function()
    local lspconfig = require("lspconfig")
    lspconfig.elixirls.setup({
      cmd = { path_to_elixirls },
      -- settings = {
      --   elixirLS = {
      --     dialyzerEnabled = false,
      --     fetchDeps = false,
      --     enableTestLenses = true,
      --   }
      -- }
    })
  end
}
```
This config currently works, but what you said about `$PATH` means I could simplify it to `cmd = { "elixir-ls" }`, assuming that is indeed resolved in my `$PATH`?

**Reply by LostInTheLogs** [reply]
if you modified `cmd` with the full path, then it doesn't actually need to be in $PATH, it should check `vim.fn.executable(cmd[1])`, so in your case you can try 
`:lua =vim.fn.executable("path_to_elixirls")` if that returns 1 then it should work

**Reply by simpers** [reply]
> if you modified `cmd` with the full path, then it doesn't actually need to be in $PATH, it should check `vim.fn.executable(cmd[1])`, so in your case you can try `:lua =vim.fn.executable("path_to_elixirls")` if that returns 1 then it should work

Thanks! That was a sanity check on my part to see how nvim resolved stuff.

I did make it available under `$PATH` anyway :)
